### PR TITLE
[ROCm] Stop hipifying native/cudnn/ and native/quantized/cudnn/ files

### PR DESCRIPTION
**Pull request by rkayaith**
This PR adds hipDNN as a supported backend for SDPA.

The approach taken here is to re-use the existing `CUDNN_ATTENTION` backend, adding support for it by routing through hipDNN when compiled for ROCm. i.e. `torch.nn.attention.SDPBackend.CUDNN_ATTENTION` and `aten::_scaled_dot_product_cudnn_attention` now work on ROCm.

The primary change here is adding `cudnn/hip/MHA.cpp`, which is a "fork" of `cudnn/MHA.cpp`, modified to use hipDNN instead of cuDNN. There's various differences between the implementations that made it simpler to just fork the entire file rather than trying to keep both implementations in the same file/rely on hipify for translation:
- various minor API differences between cuDNN/hipDNN which made it difficult to re-use code directly
- differences in feature support: hipDNN doesn't support nested tensors which simplifies a lot of code, however it *is* more flexible in other aspects e.g. dtypes. 

Additionally, since hipDNN provides an API for querying engine support for a graph, during backend selection `sdp::can_use_cudnn_attention` calls the newly added `at::native::check_cudnn_sdpa_support`, which constructs the hipDNN graph for both forwards and backwards (if potentially needed) to query support. This is *not* cached, as it's assumed construction + querying support is implemented efficiently. The sequence of hipDNN calls:
```
                SELECTION
                ─────────
       sdp::can_use_cudnn_attention()
                    │
                    ▼
    at::native::check_cudnn_sdpa_support()
                    │
                    ▼
     fe::graph::Graph::is_supported_ext()
                    │
                    ▼
                   bool
    
                EXECUTION
                ─────────
  at::native::run_cudnn_SDP_{fprop,bprop}()
                    │
                    ▼
           MHAGraphCache lookup
                    │
               ┌────┴────┐
               ▼         ▼
             HIT:      MISS:
             reuse       │
             cached      ▼
             graph     fe::graph::Graph::validate()
               │       fe::graph::Graph::build_operation_graph()
               │       fe::graph::Graph::create_execution_plans()
               │       fe::graph::Graph::check_support()
               │       fe::graph::Graph::build_plans()
               │         │
               └────┬────┘
                    ▼
         fe::graph::Graph::execute()

```

Both backends still share the same `attention.cu` kernel (hipDNN uses the hipified version).

This is separated into a stack of PRs for easier review:
- **PR 1/3: [[ROCm] Stop hipifying native/cudnn/ and native/quantized/cudnn/ files](https://github.com/ROCm/pytorch/pull/3196) (this PR)**
    - Various cudnn files are being hipified at the moment, with no functional purpose (it just results in code that's `ifdef`'d out. This disables the hipification rules and fixes up includes/directives so the CUDA files compile cleanly on ROCM. The primary motivation here is to stop generating `cudnn/hip/MHA.cpp`; the following changes add this file back with the hipDNN backend implementation.

- PR 2/3: [[ROCm] Integrate hipDNN as an SDPA backend](https://github.com/ROCm/pytorch/pull/3197):
  - I'd recommend reviewing the individual commits here:
    - `Add aotriton.images/ to .gitignore` - NFC change
    - `Extract compute_matching_strides from alloc_with_matching_layout` - NFC change
    - `Split cudnn/MHA.cpp into separate cuDNN and hipDNN files`
    - `Copy cuDNN MHA implementation verbatim into hipDNN MHA`
      - This is just to make it easier to review the next commit, as it shows the diff between cuDNN and hipDNN.
    - `[ROCm] Add hipDNN SDPA backend dispatch`
      - Primary implementation, looking at this commit will likely be the easiest way to review the hipDNN specific changes.
    - `Update tests to reflect cudnn backend being available on ROCM.`
      - `CUDNN_ATTENTION` tests can now be run on ROCm.

- PR 3/3: [[ROCm] Pass bool attention masks directly to hipDNN](https://github.com/ROCm/pytorch/pull/3198)
    - hipDNN supports boolean attention masks more efficiently than float masks. This skips the float->bool conversion that's normally done, and passes the original mask directly to hipDNN.

Open questions for reviewers:
- Reusing the `CUDNN_ATTENTION` backend was done due to API similarities, though this could potentially be confusing. Would it be preferrable to completely separate hipDNN by adding a new backend?
- Would additional code-sharing between the cuDNN and hipDNN `MHA.cpp` files be recommended?

### CODE REVIEW

File: aten/src/ATen/native/cudnn/RNN.cpp [hunk 1]
```diff
@@ -113,6 +110,9 @@ Tensor _cudnn_init_dropout_state(
 
 #else // AT_CUDNN_ENABLED()
 
+#include <ATen/cuda/CUDAEvent.h>
```
Comment (zjgarvey) [inline aten/src/ATen/native/cudnn/RNN.cpp:113]: Is this to save some linking time for non-cudnn builds?

File: aten/src/ATen/CMakeLists.txt [hunk 2]
```diff
@@ -644,15 +644,15 @@ if(USE_ROCM)
     ${native_sparse_hip_cpp}
     ${native_quantized_hip_cpp}
     ${native_transformers_hip_cpp}
-    ${native_quantized_cudnn_hip_cpp}
+    ${native_quantized_cudnn_cpp}
```
Comment (zjgarvey) [inline aten/src/ATen/CMakeLists.txt:647]: What are we currently using from these sources? Some hipified cudnn kernels? Or nothing right now? Should we remove instead?

**Review by zjgarvey — COMMENTED** [review]
Some questions. Definitely want an eye from @jeffdaily on some of these changes, too.

File: aten/src/ATen/native/cudnn/RNN.cpp (hunk 1 above)
Comment (rkayaith) [reply aten/src/ATen/native/cudnn/RNN.cpp:113]: It's needed for things to compile. The headers don't work directly on ROCM. Previously hipify was rewriting the includes to point to a a rocm version of the header, which was how the hipified version was able to compile. 

Since the non-cudnn paths don't even need the headers, I just pushed them down into the cudnn-specific sections, which lets us directly compile the original file on ROCM.

File: aten/src/ATen/CMakeLists.txt (hunk 2 above)
Comment (rkayaith) [reply aten/src/ATen/CMakeLists.txt:647]: In some cases there's an implementation of the function that just fails at runtime with a "not compiled with cudnn" error, e.g.
https://github.com/ROCm/pytorch/blob/fe50fbc0508906dcdcf0bd2b487580278c268a2d/aten/src/ATen/native/cudnn/RNN.cpp#L35-L47

I didn't look into why/if that's really necessary, just opting to preserve the existing behaviour here.

**Review by zjgarvey — APPROVED** [review]
Looks good to me, but might be good to have @jeffdaily take a look to see if there is anything that might be problematic about the hipify changes.
